src/main/java/gregtech/common/covers/CoverFluidRegulator.java

```diff
@@ -73,32 +74,129 @@ protected int doTransferExact(int transferLimit, IFluidHandler sourceHandler, IF
+
+                // This Fluid can be drained and filled, so let's move the most that will actually work.
+                int fluidToMove = Math.min(drainable, fillResult);
+                FluidStack drainedActual = sourceHandler.drain(copyFluidStackWithAmount(fluidStack, fluidToMove), true);
+
+                // Account for potential error states from the drain
+                if(drainedActual == null)
+                    throw new RuntimeException("Misbehaving fluid container: drain produced null after simulation succeeded");
+
+                if(!fluidStack.equals(drainedActual))
+                    throw new RuntimeException("Misbehaving fluid container: drain produced a different fluid than the simulation");
+
+                if(drainedActual.amount != fluidToMove)
+                    throw new RuntimeException(new FormattedMessage(
+                        "Misbehaving fluid container: drain expected: {}, actual: {}",
+                        fluidToMove,
+                        drainedActual.amount).getFormattedMessage());
+
+
+                // Perform Fill
+                int filledActual = destHandler.fill(copyFluidStackWithAmount(fluidStack, fluidToMove), true);
+
+                // Account for potential error states from the fill
+                if(filledActual != fluidToMove)
+                    throw new RuntimeException(new FormattedMessage(
+                        "Misbehaving fluid container: fill expected: {}, actual: {}",
+                        fluidToMove,
+                        filledActual).getFormattedMessage());
+
+                // update the transferred amount
+                transferred += fluidToMove;
             }
-            if (fluidLeftToTransfer == 0) break;
         }
-        return transferLimit - fluidLeftToTransfer;
+
+        return transferred;
+    }
+
+    /**
+     * Copies a FluidStack and sets its amount to the specified value.
+     *
+     * @param fs     the original fluid stack to copy
+     * @param amount the amount to set the copied FluidStack to
+     * @return the copied FluidStack with the specified amount
+     */
+    private static FluidStack copyFluidStackWithAmount(FluidStack fs, int amount) {
+        FluidStack fs2 = fs.copy();
+        fs2.amount = amount;
+        return fs2;
+    }
+
+    private Map<FluidStack, Integer> collectDistinctFluids(IFluidHandler handler,
+                                                     Predicate<IFluidTankProperties> tankTypeFilter,
+                                                     Predicate<FluidStack> fluidTypeFilter) {
+
+        final Map<FluidStack, Integer> summedFluids = new HashMap<>();
+        Arrays.stream(handler.getTankProperties())
+              .filter(tankTypeFilter)
+              .map(IFluidTankProperties::getContents)
+              .filter(Objects::nonNull)
+              .filter(fluidTypeFilter)
+              .filter(Objects::nonNull) // objects can and have been null here, shut up IntelliJ
```

Comment:
filter(Objects::nonNull) is being called twice. guess you really hate your nulls (:

Reply:
While testing the code I encountered nulls there. Not sure how exactly since I agree with IntelliJ, it shouldn't have been possible at that point. It's possible the accompanying test case was set up wrong at the time and subsequently fixed; that was one of the first methods I wrote while working on this.

Reply:
Do you need first nonNull filter shouldn't last one catch everything?

Reply:
As I told Proto 15 days ago, I figured I wouldn't have needed it but I got a NPE without it. I could try it again and see if it was an unrelated problem.